### PR TITLE
Extract LINUX_VERSION_CODE from running kernel via vDSO.

### LINKED ISSUE #206: Error loading program, with no details (kern version)
This is due to bpftrace being compiled on one kernel, then run on another. Errors like this:

```
# /usr/local/bin/bpftrace -e 'kprobe:do_nanosleep { printf("hi\n"); }'
Attaching 1 probe...
Error loading program: kprobe:do_nanosleep (try -v)
```

with -v:

```
# /usr/local/bin/bpftrace -v -e 'kprobe:do_nanosleep { printf("hi\n"); }'
Attaching 1 probe...

Error log: 

Error loading program: kprobe:do_nanosleep
```

with strace:

```
# strace -fe bpf /usr/local/bin/bpftrace -v -e 'kprobe:do_nanosleep { printf("hi\n"); }'
bpf(BPF_MAP_CREATE, {map_type=BPF_MAP_TYPE_PERF_EVENT_ARRAY, key_size=4, value_size=4, max_entries=2, map_flags=0, inner_map_fd=0, ...}, 72) = 3
Attaching 1 probe...
bpf(BPF_MAP_UPDATE_ELEM, {map_fd=3, key=0x7fff100009c0, value=0x7fff100009c4, flags=BPF_ANY}, 72) = 0
bpf(BPF_MAP_UPDATE_ELEM, {map_fd=3, key=0x7fff100009c0, value=0x7fff100009c4, flags=BPF_ANY}, 72) = 0
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=15, insns=0x7f51fe40d000, license="GPL", log_level=1, log_size=102400, log_buf=0x7fff0ffe78a0, kern_version=266002, prog_flags=0, ...}, 72) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=15, insns=0x7f51fe40d000, license="GPL", log_level=1, log_size=102400, log_buf=0x7fff0ffe78a0, kern_version=266002, prog_flags=0}, 72) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=15, insns=0x7f51fe40d000, license="GPL", log_level=1, log_size=102400, log_buf=0x7fff0ffe78a0, kern_version=265984, prog_flags=0, ...}, 72) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=15, insns=0x7f51fe40d000, license="GPL", log_level=1, log_size=102400, log_buf=0x7fff0ffe78a0, kern_version=265984, prog_flags=0}, 72) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=15, insns=0x7f51fe40d000, license="GPL", log_level=1, log_size=102400, log_buf=0x7fff0ffe78a0, kern_version=0, prog_flags=0, ...}, 72) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=15, insns=0x7f51fe40d000, license="GPL", log_level=1, log_size=102400, log_buf=0x7fff0ffe78a0, kern_version=0, prog_flags=0}, 72) = -1 EINVAL (Invalid argument)

Error log: 

Error loading program: kprobe:do_nanosleep
+++ exited with 255 +++
```

See the kern_version numbers above (266002, 265984, 0). Now compare them to a working bcc invocation, which uses 266001:

```
# strace -fe bpf /usr/share/bcc/tools/execsnoop 
bpf(BPF_MAP_CREATE, {map_type=BPF_MAP_TYPE_PERF_EVENT_ARRAY, key_size=4, value_size=4, max_entries=2, map_flags=0, inner_map_fd=0, ...}, 72) = 3
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=511, insns=0x7f0db9d99000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=266001, prog_flags=0, ...}, 72) = 4
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=60, insns=0x7f0db9d9a7d0, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=266001, prog_flags=0, ...}, 72) = 6
PCOMM            PID    PPID   RET ARGS
bpf(BPF_MAP_UPDATE_ELEM, {map_fd=3, key=0x7f0db3f6f880, value=0x7f0db3f6f910, flags=BPF_ANY}, 72) = 0
bpf(BPF_MAP_UPDATE_ELEM, {map_fd=3, key=0x7f0db3f6f910, value=0x7f0db3f6f880, flags=BPF_ANY}, 72) = 0
```

We should switch to bcc's scheme for determining the kernel version.

And add a better error message.

### LINKED ISSUE #274: Kprobe LINUX_VERSION_CODE issue
I can see what's broken, but I'm not sure how it's supposed to work.

- Start with Ubuntu 18.04 default kernel, install dbgsyms and headers
- Install bpftrace deps per INSTALL.md
- Build bpftrace, observe general success
- Attempt to use 'cgroup' variable, notice it was introduced after 4.15, programs on earlier kernels fail to load.
- Go ferreting for newer Ubuntu kernel, find https://launchpad.net/~canonical-kernel-team/+archive/ubuntu/ppa , install 4.18.0-12-generic from there
- *Discover bpftrace no longer functions*
- Go digging, notice LINUX_VERSION_CODE check in `kernel/bpf/syscall.c`
- Notice bpftrace is already basically bruteforcing this field before failing
- Notice none of the LINUX_VERSION_CODE tried match the running kernel per the `#define` in the associated -headers package.
- Hack the value from the -headers package into bpftrace, observe success.

In `attached_kprobe.cpp`, we see it first tries:

- LINUX_VERSION_CODE as defined in the translation unit, which I think comes from `/usr/include/linux/version.h`.
- Regexing the equivalent constant out of `/usr/include/linux/version.h`
- Falling back to attempting to generate the constant via `uname`, which somehow fails here.

After hard-wiring the constant:

````
sudo strace -febpf bpftrace -e 'BEGIN { printf("hi\n"); exit(); }'
bpf(BPF_MAP_CREATE, {map_type=BPF_MAP_TYPE_PERF_EVENT_ARRAY, key_size=4, value_size=4, max_entries=8, map_flags=0, inner_map_fd=0, ...}, 96) = 3
Attaching 1 probe...
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=27, insns=0x7fbc8f58c000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=266769, prog_flags=0, ...}, 96) = 5
````

Prior to hard-wiring:

````
build] sudo strace -febpf ./src/bpftrace -e 'BEGIN { printf("hi\n"); exit(); }'
bpf(BPF_MAP_CREATE, {map_type=BPF_MAP_TYPE_PERF_EVENT_ARRAY, key_size=4, value_size=4, max_entries=8, map_flags=0, inner_map_fd=0, ...}, 96) = 3
Attaching 1 probe...
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=27, insns=0x7f49f9a55000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=266002, prog_flags=0, ...}, 96) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=27, insns=0x7f49f9a55000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=266002, prog_flags=0}, 96) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=27, insns=0x7f49f9a55000, license="GPL", log_level=1, log_size=102400, log_buf=0x7fffa552be40, kern_version=266002, prog_flags=0}, 96) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=27, insns=0x7f49f9a55000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=266752, prog_flags=0, ...}, 96) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=27, insns=0x7f49f9a55000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=266752, prog_flags=0}, 96) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=27, insns=0x7f49f9a55000, license="GPL", log_level=1, log_size=102400, log_buf=0x7fffa552be40, kern_version=266752, prog_flags=0}, 96) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=27, insns=0x7f49f9a55000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=266002, prog_flags=0, ...}, 96) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=27, insns=0x7f49f9a55000, license="GPL", log_level=0, log_size=0, log_buf=0, kern_version=266002, prog_flags=0}, 96) = -1 EINVAL (Invalid argument)
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_KPROBE, insn_cnt=27, insns=0x7f49f9a55000, license="GPL", log_level=1, log_size=102400, log_buf=0x7fffa552be40, kern_version=266002, prog_flags=0}, 96) = -1 EINVAL (Invalid argument)
````

uname -r output:

````
4.18.0-12-generic
````

Is this possible a Fedora vs. Debian directory layout thing?

### PR DESCRIPTION
**Pull request by dw**
This method avoids runtime header files or regexes, and the vDSO
structure should remain stable over time. Fall back to the compile time
constant if search fails.

Closes #206 
Closes #274 

See comments at end of #274 leading to this approach

### CODE REVIEW

**Comment by dw** [comment]
I've pushed another version that restores the old strategies, but augments the first with the new method.

**Comment by danobi** [comment]
I tested your patch locally and it didn't seem to fix the error:
```
$ sudo ./build/src/bpftrace -e 'tracepoint:syscalls:sys_enter_read /comm == "a.out"/ { printf("hit\n") }'                                                                                   
chdir(/lib/modules/4.18.9-200.fc28.x86_64/build): No such file or directory
definitions.h:9:3: error: unknown type name 'u64'
definitions.h:11:3: error: unknown type name 'size_t'
Attaching 1 probe...
Error loading program: tracepoint:syscalls:sys_enter_read (try -v)

$ sudo ./build/src/bpftrace -e 'tracepoint:syscalls:sys_enter_read /comm == "a.out"/ { printf("hit %d\n", args->fd) }'
chdir(/lib/modules/4.18.9-200.fc28.x86_64/build): No such file or directory
definitions.h:9:3: error: unknown type name 'u64'
definitions.h:11:3: error: unknown type name 'size_t'
Attaching 1 probe...
^C
```

I'm not fully convinced I have the same issue. I'll keep debugging too.

**Comment by danobi** [comment]
Looks like I'm hitting #198 . Please ignore me :)

**Comment by dw** [comment]
I haven't done C in quite a few years, as best I can tell the integer types / arithmetic is fine, but I always sucked at this stuff even when I did C regularly :)  Stylistically also wasn't sure about 'auto' use either

File: src/attached_probe.cpp [hunk 1]
```diff
@@ -193,12 +197,67 @@ uint64_t AttachedProbe::offset() const
...
   {
     case 0:
-      return LINUX_VERSION_CODE;
+    {
+      // Fetch LINUX_VERSION_CODE from the vDSO .note section, falling back on
+      // the build-time constant if unavailable. This always matches the
+      // running kernel, but is not supported on arm32.
+      unsigned code = 0;
+      unsigned long base = getauxval(AT_SYSINFO_EHDR);
+      if (base && !memcmp(reinterpret_cast<void *>(base), ELFMAG, 4))
```
Comment (danobi) [inline src/attached_probe.cpp]: I think `static_cast` should work here:

https://stackoverflow.com/questions/310451/should-i-use-static-cast-or-reinterpret-cast-when-casting-a-void-to-whatever

File: src/attached_probe.cpp [hunk 2]
```diff
@@ -193,12 +197,67 @@ uint64_t AttachedProbe::offset() const
...
+
+  for (int i = 0; i < ehdr->e_shnum; i++)
+  {
+    auto shdr = reinterpret_cast<const ElfW(Shdr) *>(
+      base + ehdr->e_shoff + (i * ehdr->e_shentsize)
+    );
+
+    if (shdr->sh_type == SHT_NOTE)
+    {
+      auto p = reinterpret_cast<const char *>(base + shdr->sh_offset);
```
Comment (danobi) [inline src/attached_probe.cpp]: Maybe `ptr`? More descriptive IMO.

File: src/attached_probe.cpp [hunk 3]
```diff
@@ -193,12 +197,67 @@ uint64_t AttachedProbe::offset() const
...
+      auto e = p + shdr->sh_size;
```
Comment (danobi) [inline src/attached_probe.cpp]: `end` instead of `e`?

Comment (dw) [reply src/attached_probe.cpp]: Also changed :)

File: src/attached_probe.cpp (hunk 2 above)
Comment (dw) [reply src/attached_probe.cpp]: Changed, thanks.

**Comment by brendangregg** [comment]
thanks @dw and @danobi. Looks good, and works for me.

**Pull request merged**